test/libc/stdio/fgets_interrupt_test.c

```diff
@@ -0,0 +1,138 @@
+│ THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL                │
+│ WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED                │
+│ WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE             │
+│ AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL         │
+│ DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR        │
+│ PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER               │
+│ TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR             │
+│ PERFORMANCE OF THIS SOFTWARE.                                                │
+╚─────────────────────────────────────────────────────────────────────────────*/
+#include "libc/isystem/errno.h"
+#include "libc/isystem/sched.h"
+#include "libc/isystem/signal.h"
+#include "libc/isystem/stddef.h"
+#include "libc/isystem/unistd.h"
+#include "libc/stdio/stdio.h"
+#include "libc/testlib/testlib.h"
+
+#define MY_TEST_STRING_1 "He"
+#define MY_TEST_STRING_2 "llo world!"
+
+char buf[20] = {0};
+int pipes[2];
+int pid;
+int got_sigusr1 = 0;
+
+// -=- these get called for each test ------------------------------------------
+
+void sigusr1_handler(int) {
+  got_sigusr1 = 1;
+}
+
+void write_pipe(int send_signal_before_end) {
+  // Set up pipe for writing
+  close(pipes[0]);
+  FILE *stream = fdopen(pipes[1], "w");
+
+  // Start writing the first part of the stream
+  fputs(MY_TEST_STRING_1, stream);
+
+  // Send SIGUSR1 to parent (if we're currently testing that)
+  if (send_signal_before_end) {
+    kill(getppid(), SIGUSR1);
+  }
+
+  // Send rest of stream
+  fputs(MY_TEST_STRING_2, stream);
+  // Close stream - this will cause the parent's fgets to end
+  fclose(stream);
+}
+
+void read_pipe() {
+  // Set up pipe for reading
+  close(pipes[1]);
+  FILE *stream = fdopen(pipes[0], "r");
+
+  // Read with fgets
+  fgets(buf, 20, stream);
+
+  // Tidy up
+  fclose(stream);
+}
+
+// -=- these set up the tests --------------------------------------------------
+
+void SetUpOnce(void) {
+  cpu_set_t set;
+  CPU_ZERO(&set);
+  CPU_SET(1, &set);
+  if (sched_setaffinity(0, sizeof set, &set) == -1) {
+    perror("sched_setaffinity");
+    fprintf(stderr, "single core affinity is needed for test reliability\n");
+    _exit(1);
+  }
+}
+
+void setup_signal_and_pipe(uint64_t sa_flags) {
+  // Set up SIGUSR1 handler
+  struct sigaction sa = {.sa_handler = sigusr1_handler, .sa_flags = sa_flags};
+  if (sigaction(SIGUSR1, &sa, NULL) == -1) {
+    perror("sigaction");
+    _exit(1);
+  }
+  got_sigusr1 = 0;
+
+  // Set up pipe between parent and child
+  if (pipe(pipes) == -1) {
+    perror("pipe");
+    _exit(1);
+  }
+}
+
+// -=- these are the tests -----------------------------------------------------
+
+TEST(fgets_eintr, testThatFgetsReadsFromPipeNormally) {
+  setup_signal_and_pipe(0);  // 0 = no SA_RESTART
+  ASSERT_NE(-1, (pid = fork()));
```

Comment:
Consider rather than using `fork`,  `exit`, etc directly  use `SPAWN`, `PARENT`, and `WAIT` from `#include "libc/testlib/subprocess.h"` as a standard way to `fork`  and wait on children hooked in with the test library. See [`test/libc/sock/recvfrom_test.c` ](https://github.com/jart/cosmopolitan/blob/master/test/libc/sock/recvfrom_test.c) for an example.

On `testThatTheSignalInterruptsFgets` have the child `EXPECT_SYS(0, 0, sigignore(SIGPIPE));` so that it exits successfully.